문의 글 삭제

sudago opened the pull request:
## What is this PR? 👓
- https://github.com/jazz-meet/jazz-meet/issues/139

## Key changes 🔑
- 문의를 soft delete 하게 되어 문의 상태에 DELETED(enum InquiryStatus) 를 추가했습니다.
- 비밀번호 암호화 방식을 변경했습니다. 
    - EncyptPasswordEncoder -> EncryptPasswordEncoder 로 교체

## To reviewers 👋
### 비밀번호 암호화 방식을 변경한 이유 상세
-  기존에 직접 만들었던 EncryptPasswordEncoder은 salt(=Secret Key, 비밀번호 뒤에 붙은 임의의 문자열)의 값이 고정된다는 문제가 있었습니다.
    - 이는 동일한 비밀번호가 들어왔을 때, 암호화된 값이 동일한 값으로 출력되어 보안이 저하되게 됩니다.
    - 이 때, 어떤 나쁜 사람이 DB를 해킹한다면, salt(Secret Key) 값을 유추해내기 쉬워집니다. (레인보우 어쩌구 문제 발생..)
- spring security에서 제공해주는 BcryptPasswordEncoder로 해결하게 되었습니다.
    - BcryptPasswordEncode는 salt를 랜덤으로 생성해내, 같은 비밀번호라도 다른 값으로 인코딩 되게 되어 해당 문제를 해결해줍니다
    - 또한 요청으로 오는 비밀번호와, DB에 인코딩(암호화) 된 비밀번호가 일치하는지에 대한 검사도 BcryptPasswordEncoder가 제공해줍니다.

- spring security가 제공해주는 모듈을 사용하게 된다면 security가 멋대로 우리 어플리케이션에 대한 보안 작업을 해버리진 않을까 고민했는데, security-starter를 사용하면 그렇게 해주는 것 같고, security-core만 의존성에 추가하게 되면 그렇게 되지 않으면서 원하는 모듈을 사용할 수 있게 되는 것 같아 security-core를 의존성에 추가하게 되었습니다.

sudago committed:
chore: 불필요한 생성자 제거

sudago committed:
fix: 문의 상태에 삭제됨 추가 및 조회 조건 수정

sudago committed:
fix: 테이블 생성 시 DB default 값 추가

sudago committed:
feat: 문의 글 삭제 기능 구현

sudago committed:
fix: BcryptPasswordEncoder로 암호화 모듈 교체

직접 만든 암호화 클래스에서 spring security에서 제공해주는 BcryptPasswordEncoder로 암호화 모듈 교체.
이유: 직접 만든 클래스는 고정된 Salt 값을 사용하는 반면, BcryptPasswordEncoder에서 salt 값은 계속 바뀌기 때문에 보안적으로 더 우수하다고 할 수 있다.

sudago committed:
feat: 문의 삭제 시 예외 추가

이미 삭제 된 문의에 대한 예외 추가

sudago committed:
chore: 삭제 메서드 네이밍 변경

sudago committed:
test: 문의 글 삭제 테스트

sudago committed:
chore: 에러코드 이름 변경

sudago committed:
fix: 문의 상세 조회 request 조건 수정

sudago committed:
Merge pull request #151 from jazz-meet/dev

dev 브랜치 be/feat/#139-문의 글 삭제 브랜치로 merge

sudago committed:
Merge remote-tracking branch 'origin/be/feat/#139-문의-글-삭제' into be/feat/#139-문의-글-삭제

sudago committed:
test: venue 테스트 수정

sudago committed:
fix: 불필요한 required=false 제거

request로 category, page가 요청되지 않을 때에도, default 값을 설정하면 요청 온 것 처럼 설정이 된다.
가능한 예시1) api/inquiries/page=
가능한 예시2) api/inquiries/

chunghye98 commented on be/src/main/java/kr/codesquad/jazzmeet/global/config/EncoderConfig.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+
+@Configuration
+public class EncoderConfig {
+
+	@Bean
+	public BCryptPasswordEncoder bCryptPasswordEncoder() {
+		return new BCryptPasswordEncoder();
+	}
+}
```
이렇게 빈으로 등록해서 사용하는 이유가 무엇일까요?

sudago replied:
스프링이 해당 빈 객체를 관리해주기 때문에 어느 클래스든 필드에 작성하면 주입 받아 사용하기 쉬워집니다.
= 쓰기 편하려고 &  BCryptPasswordEncoder.encode() 하면 에러남

chunghye98 commented on be/src/main/java/kr/codesquad/jazzmeet/inquiry/dto/request/InquiryDeleteRequest.java:
```diff
@@ -0,0 +1,6 @@
+package kr.codesquad.jazzmeet.inquiry.dto.request;
+
+public record InquiryDeleteRequest(
+	String password
+) {
+}
```
password에 대한 valid 조건이 있으면 더 좋지 않을까요?

sudago replied:
글자 수 제한 추가해놨습니다~

chunghye98 commented on be/src/main/java/kr/codesquad/jazzmeet/inquiry/entity/Answer.java:
```diff
@@ -36,10 +36,10 @@ public class Answer {
 	@CreationTimestamp
-	@Column(nullable = false)
+	@Column(nullable = false, columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP")
 	private LocalDateTime createdAt;
 	@UpdateTimestamp
-	@Column(nullable = false)
+	@Column(nullable = false, columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP")
```
👍

chunghye98 commented on be/src/main/java/kr/codesquad/jazzmeet/inquiry/service/InquiryService.java:
```diff
@@ -71,12 +73,35 @@ public InquiryDetailResponse getInquiryDetail(Long inquiryId) {
 
+	@Transactional
+	public void delete(Long inquiryId, InquiryDeleteRequest request) {
+		Inquiry inquiry = inquiryRepository.findById(inquiryId)
+			.orElseThrow(() -> new CustomException(InquiryErrorCode.NOT_FOUND_INQUIRY));
```
이 로직은 분리하면 나중에 재사용하기 좋을 것 같아요

sudago replied:
넴 수정해서 반영해놨습니다!

chunghye98 commented on be/src/test/java/kr/codesquad/jazzmeet/venue/service/VenueServiceTest.java:
```diff
@@ -409,6 +409,7 @@ void searchVenueListContainsWordInAddress() {
 		assertThat(venueSearchResponse.currentPage()).isEqualTo(page);
+		assertThat(venueSearchResponse.maxPage()).isEqualTo(1);
 		assertThat(venueSearchResponse.venues()).extracting(VenueSearch::address)
 			.doesNotContain(venue3.getRoadNameAddress())
```
doesnotContain()은 어떤 메소드인가요?

sudago replied:
말 그대로 해당 객체는 반환되지 않아야 하기 때문에 반환되지 않는지에 대한 검사입니다~!

jsjseop replied:
오호👍

sudago committed:
fix: valid로 글자 수 제한 추가

sudago committed:
refactor: 메서드 분리

sudago committed:
refactor: 메서드 역할 분리

재사용과 직관성, 역할 분리를 명확하게 하기 위해 상위 메서드에서 직접 getStatus로 상태를 꺼내 argument 넣는 것으로 변경, 하위 메서드 Parameter 리팩터링

sudago committed:
chore: 반환 타입 변경

sudago committed:
fix: 문의 조회 목록 최신순 정렬

jsjseop reviewed (approved):
코드가 깔끔하고 좋네요👍
